mindsdb/utilities/cache.py

```diff
@@ -1,121 +1,230 @@
+"""
+How to use it:
+
+    from mindsdb.utilities.cache import get_cache, dataframe_checksum, json_checksum
+
+    # namespace of cache
+    cache = get_cache('predict')
+
+    key = dataframe_checksum(df) # or json_checksum, depends on object type
+    df_predict = cache(key)
+
+    if df_predict is None:
+        # no cache, save it
+        df_predict = predictor.predict(df)
+        cache.set(key, df_predict)
+
+
+
+Configuration:
+
+- max_size size of cache in count of records, default is 50
+- serializer, module for serialization, default is dill
+
+It can be set via:
+- get_cache function:
+    cache = get_cache('predict', max_size=2)
+- using specific cache class:
+    cache = FileCache('predict', max_size=2)
+- using mindsdb config file:
+    "cache": {
+        "type": "redis",
+        "max_size": 2
+    }
+
+Cache engines:
+
+Can be specified in mindsdb config json. Possible values:
+- local - for FileCache, default
+- redis - for RedisCache
+By default is used local redis server. You can specify
+    "cache": {
+        "type": "redis",
+        "connection": {
+            "host": "127.0.0.1",
+            "port": 6379
+        }
+    }
+
+How to test:
+
+    env PYTHONPATH=./ pytest tests/unit/test_cache.py
+
+"""
+
 import os
-import shelve
+import time
+from abc import ABC
+from pathlib import Path
+import hashlib
 import json
-from abc import ABC, abstractmethod
-
+import dill
+import pandas as pd
 import walrus
+
 from mindsdb.utilities.config import Config
 
 CONFIG = Config()
 
 
+def dataframe_checksum(df: pd.DataFrame):
+    checksum = str_checksum(df.to_json())
+    return checksum
+
+
+def json_checksum(obj: [dict, list]):
+    checksum = str_checksum(json.dumps(obj))
+    return checksum
+
+
+def str_checksum(obj: str):
+    checksum = hashlib.sha256(obj.encode()).hexdigest()
+    return checksum
+
+
 class BaseCache(ABC):
-    def __init__(self):
+    def __init__(self, max_size=None, serializer=None):
         self.config = Config()
+        if max_size is None:
+            max_size = self.config["cache"].get("max_size", 50)
+        self.max_size = max_size
+        if serializer is None:
+            self.serializer = self.config["cache"].get('serializer', dill)
```

Comment:
config is json file - there may be only text value 'dill' in it

Reply:
Changed to string